sdk/monitor/Azure.Monitor.OpenTelemetry.Exporter/src/Internals/AzMonNewListExtensions.cs

```diff
@@ -15,28 +16,83 @@ internal static class AzMonNewListExtensions
     {
         try
         {
-            var serverAddress = AzMonList.GetTagValue(ref tagObjects, SemanticConventions.AttributeServerAddress)?.ToString();
-            if (serverAddress != null)
+            var host = AzMonList.GetTagValue(ref tagObjects, SemanticConventions.AttributeServerAddress)?.ToString();
+            if (host != null)
             {
-                UriBuilder uriBuilder = new()
-                {
-                    Scheme = AzMonList.GetTagValue(ref tagObjects, SemanticConventions.AttributeUrlScheme)?.ToString(),
-                    Host = serverAddress,
-                    Path = AzMonList.GetTagValue(ref tagObjects, SemanticConventions.AttributeUrlPath)?.ToString(),
-                    Query = AzMonList.GetTagValue(ref tagObjects, SemanticConventions.AttributeUrlQuery)?.ToString()
-                };
+                var requestUrlTagObjects = AzMonList.GetTagValues(ref tagObjects, SemanticConventions.AttributeUrlScheme, SemanticConventions.AttributeUrlPath, SemanticConventions.AttributeUrlQuery, SemanticConventions.AttributeServerPort);
+                var scheme = requestUrlTagObjects[0]?.ToString();
+                var path = requestUrlTagObjects[1]?.ToString();
+                var queryString = requestUrlTagObjects[2]?.ToString();
+                var queryStringLength = queryString?.Length;
+                var isNonDefaultPort = TryGetNonDefaultPort(requestUrlTagObjects[3]?.ToString(), out string? port);
+                var length = (scheme?.Length ?? 0) + (scheme?.Length > 0 ? Uri.SchemeDelimiter.Length : 0) + host.Length + (port?.Length > 0 ? 1 : 0) + (port?.Length ?? 0) + (path?.Length ?? 1) + (queryString?.Length > 0 ? 1 : 0) + (queryString?.Length ?? 0);
+
+                var urlStringBuilder = new System.Text.StringBuilder(length)
+                    .Append(scheme)
+                    .Append(scheme != null ? Uri.SchemeDelimiter : null)
+                    .Append(host)
+                    .Append(isNonDefaultPort ? $":{port}" : null)
+                    .Append(string.IsNullOrEmpty(path) ? "/" : path)
+                    .Append(queryString != null ? $"?{queryString}" : null);
```

Comment:
`$"?{queryString}"`
Do we need to add `?` here?

Reply:
`uri.query` does not include `?` in the spec. Hence added manually here. I re-checked the spec now, in one to the example `?` is  a part of it. We need to drive clarity here.

@TimothyMothra  Did you add `?` to the `uri.query` in AspNetCoreInstrumetation or HttpClientInstrumentation?

Reply:
I think it should be as per [this](https://www.rfc-editor.org/rfc/rfc3986#section-3.4). would be good to test if ASP.NET Core and ASP.NET adds it

Reply:
I'm not seeing any unit tests that check for the query string.
We should include tests in the Exporter

Reply:
Current PR covers that test - `GetNewRequestUrl_ReturnsCorrectUrl`